packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart

```diff
@@ -134,22 +227,77 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
     throw Exception('Implement this when needed');
   }
 
+  double getLogHeight(int index) {
+    final entry = _logs.elementAt(index);
+    final cachedHeight = entry.height;
+    if (cachedHeight != null) return cachedHeight;
+    final height = LoggingTableRow.calculateRowHeight(
+      entry.log,
+      _tableWidth,
+    );
+    entry.height = height;
+    return height;
+  }
+
   /// Get the height of a filtered Log at [index].
   double getFilteredLogHeight(int index) {
-    final cachedHeight = cachedHeights[index];
+    final filteredLog = _filteredLogs.elementAt(index);
+    final cachedHeight = filteredLog.logEntry.height;
     if (cachedHeight != null) return cachedHeight;
 
-    return cachedHeights[index] = LoggingTableRow.calculateRowHeight(
-      _logs.elementAt(index),
+    final height = LoggingTableRow.calculateRowHeight(
+      filteredLog.logEntry.log,
       _tableWidth,
     );
+    filteredLog.logEntry.height = height;
+    return height;
   }
 
   Future<bool> _preFetchRowHeights() async {
     final didComplete = await _worker.doWork(_logs.length);
     if (didComplete) {
       _cacheLoadProgress.value = null;
     }
+    _recalculateOffsets();
     return didComplete;
   }
+
+  /// The toggle filters available for the Logging screen.
+  @override
+  List<ToggleFilter<LogDataV2>> createToggleFilters() => [
+        if (serviceConnection.serviceManager.connectedApp?.isFlutterAppNow ??
+            true) ...[
+          ToggleFilter<LogDataV2>(
+            name: 'Hide verbose Flutter framework logs (initialization, frame '
+                'times, image sizes)',
+            includeCallback: (log) => !_verboseFlutterFrameworkLogKinds
+                .any((kind) => kind.caseInsensitiveEquals(log.kind)),
+            enabledByDefault: true,
+          ),
+          ToggleFilter<LogDataV2>(
+            name: 'Hide verbose Flutter service logs (service extension state '
+                'changes)',
+            includeCallback: (log) => !_verboseFlutterServiceLogKinds
+                .any((kind) => kind.caseInsensitiveEquals(log.kind)),
+            enabledByDefault: true,
+          ),
+        ],
+        ToggleFilter<LogDataV2>(
+          name: 'Hide garbage collection logs',
+          includeCallback: (log) => !log.kind.caseInsensitiveEquals(_gcLogKind),
+          enabledByDefault: true,
+        ),
+      ];
+}
+
+class _LogEntry {
+  _LogEntry(this.log);
+  final LogDataV2 log;
+  double? height;
+}
+
+class _FilteredLogEntry {
+  _FilteredLogEntry(this.logEntry);
+  final _LogEntry logEntry;
+  double? offset;
```

Comment:
do we need an entire wrapper class for these? Can't we just keep a map to lookup offsets for each _LogEntry?